dataland-community-manager/src/main/kotlin/org/dataland/datalandcommunitymanager/services/BulkDataRequestManager.kt

```diff
@@ -33,43 +42,185 @@ class BulkDataRequestManager(
+
         sendBulkDataRequestInternalEmailMessage(
-            bulkDataRequest, userProvidedIdentifierToDatalandCompanyIdMapping.values.toList(), correlationId,
+            bulkDataRequest, acceptedIdentifiersToCompanyIdAndName.values.toList(), correlationId,
+        )
+
+        return createBulkDataRequests(
+            acceptedRequestCombinations = acceptedRequestCombinations,
+            acceptedIdentifiersToCompanyIdAndName = acceptedIdentifiersToCompanyIdAndName,
+            existingDatasets = existingDatasets,
+            rejectedIdentifiers = rejectedIdentifiers,
+            correlationId = correlationId,
         )
-        return buildResponseForBulkDataRequest(bulkDataRequest, rejectedIdentifiers, acceptedIdentifiers)
     }
 
-    private fun storeDataRequests(
+    private fun getValidRequestCombinations(
+        bulkDataRequest: BulkDataRequest,
+        acceptedIdentifiersToCompanyIdAndName: Map<String, CompanyIdAndName>,
+    ): List<DataMetaInformationRequest> {
+        val acceptedCompanyIdsSet: Set<String> = acceptedIdentifiersToCompanyIdAndName.values.map { it.companyId }.toSet()
+
+        return generateRequestCombinations(
+            dataTypes = bulkDataRequest.dataTypes,
+            reportingPeriods = bulkDataRequest.reportingPeriods,
+            datalandCompanyIds = acceptedCompanyIdsSet,
+        )
+    }
+
+    private fun getAlreadyExistingDatasetsResponse(
+        metaDataList: List<DataMetaInformation>,
+        userProvidedIdentifierToDatalandCompanyIdMapping: Map<String, CompanyIdAndName>,
+    ): List<DatasetsResponse> {
+        if (metaDataList.isEmpty()) {
+            return emptyList()
+        }
+
+        return metaDataList.map { metaData ->
+            val companyId = metaData.companyId
+
+            val companyMappingEntry =
+                userProvidedIdentifierToDatalandCompanyIdMapping.entries
+                    .firstOrNull { it.value.companyId == companyId }
+
+            companyMappingEntry ?: throw IllegalArgumentException("Can't match: $companyId to a user provided company identifier.")
+
+            DatasetsResponse(
+                userProvidedCompanyId = companyMappingEntry.key,
+                companyName = companyMappingEntry.value.companyName,
+                framework = metaData.dataType.toString(),
+                reportingPeriod = metaData.reportingPeriod,
+                datasetId = metaData.dataId,
+                datasetUrl = metaData.url,
+            )
+        }
+    }
+
+    private fun getAcceptedRequestCombinations(
+        validRequestCombinations: List<DataMetaInformationRequest>,
+        existingDatasets: List<DataMetaInformation>,
+    ): List<ValidSingleDataRequest> {
+        val existingCriteriaSet =
+            existingDatasets
+                .map {
+                    Triple(it.companyId, it.dataType, it.reportingPeriod)
```

Comment:
This triple is again the data dimension and should be named and used accordingly.